Add SSL/TLS articles to conceptual docs (#32194)
dotnet/docs: docs/core/extensions/sslstream-best-practices.md

---
title: TLS/SSL best practices
description: Learn best practices to using SslStream in .NET
author: rzikm
ms.author: rzikm
ms.date: 10/25/2022
---

# TLS/SSL best practices

TLS (Transport Layer Security) is a cryptographic protocol designed to secure communication between two computers over the internet. The TLS protocol is exposed in .NET via the <xref:System.Net.Security.SslStream> class.

> This article presents best practices for setting up secure communication between client and server and assumes .NET Core version 3.1 or later. For best practices for .NET Framework, see [Transport Layer Security (TLS) best practices with the .NET Framework](/dotnet/framework/network-programming/tls).

Reviewer: 3.1 is out of support as off today. While that does not matter for the recommendation I would probably not specify any version or go back all the way to 2.x. Starting with 6, we no longer call it "Core" and that makes it even more confusing IMHO.

## Selecting TLS version

While it is possible to specify the version of the TLS protocol to be used via the SslProtocols property, it is recommended to defer to the operating system settings by using <xref:System.Net.Security.SslProtocols.None> value (this is the default).

> Deferring the decision to the OS automatically uses the most recent version of TLS available and lets the application pick up changes after OS upgrades. The operating system may also prevent use of TLS versions which are no longer considered secure.

Reviewer: While this true, modifying OS is not always possible or practical. I'm wondering if we can craft some example to enforce minimal version. That can perhaps be done post handshake...? 

cc: @baronjs for any suggestions.

Reviewer 2: Our API doesn't support "minimal version", only "these versions", which makes it brittle when a new version comes online (e.g. specifying `Tls12 | Tls13` today means it ends up boxing out Tls14 in the future).

I think I agree with this saying simply "specify None, and let the OS decide"... which should call just enough attention to the option that someone could figure out they could specify specific versions.

## Selecting cipher suites

SslStream allows users to specify which cipher suites can be negotiated by the TLS handshake via the <xref:System.Net.Security.CipherSuitesPolicy> class. As with TLS versions, we recommend letting the operating system decide which are the best cipher suites to negotiate with and, therefore, we recommend avoiding the use of <xref:System.Net.Security.CipherSuitesPolicy>.

> Note that <xref:System.Net.Security.CipherSuitesPolicy> is not available on all platforms supported by .NET.

Reviewer: We may mention directly here that attempt to use it on Windows would throw...

## Specifying a Server Certificate

> When authenticating as a server, <xref:System.Net.Security.SslStream> requires an <xref:System.Security.Cryptography.X509Certificates.X509Certificate> instance to use. There are multiple ways how the server certificate can be passed to <xref:System.Net.Security.SslStream>:

Reviewer: I would also mention that the certificate should have private key. We will search through stores for it if it does not bat that is problematic IMHO. I think @bartonjs would also prefer to use `X509Certificate2` everywhere we can. 
On Windows the certificate/key must be in store or persistent storage so `schanel` can access it. While this is implementation details, I would probably mention it. 
Perhaps @bartonjs can recommend some import/create fragment for anybody loading it from PEM of pfx.

Author: > On Windows the certificate/key must be in store or persistent storage so schanel can access it.

Does it really? I thought this holds only for the intermediate certificates and the leaf certificate does not have to be in the store.

Reviewer: I was primarily thinking about https://github.com/dotnet/runtime/issues/23749
Getting them to store is one way how to fix it. When loaded from file, there is some trickery about the flags. 
There is bunch of linked issues as people seems to subtle on this regularly. We can also move it down to the debug section. I'm not sure if conceptual doc is good dumping ground for platform caveats.

- By passing a <xref:System.Net.Security.SslStreamCertificateContext> in the <xref:System.Net.Security.SslServerAuthenticationOptions.ServerCertificateContext?displayProperty=nameWithType> property

The recommended approach is to use the <xref:System.Net.Security.SslServerAuthenticationOptions.ServerCertificateContext?displayProperty=nameWithType> property. When the certificate is obtained by one of the other two ways, a <xref:System.Net.Security.SslStreamCertificateContext> instance is created internally by the <xref:System.Net.Security.SslStream> implementation. Creating a <xref:System.Net.Security.SslStreamCertificateContext> involves building an <xref:System.Security.Cryptography.X509Certificates.X509Chain> which is a CPU intensive operation. It is more efficient to create a <xref:System.Net.Security.SslStreamCertificateContext> once and reuse it for multiple <xref:System.Net.Security.SslStream> instances.

Reusing <xref:System.Net.Security.SslStreamCertificateContext> instances also enables additional features such us [TLS session resumption](https://datatracker.ietf.org/doc/html/rfc5077) on Linux servers.

## Custom X509Certificate validation

> There are certain scenarios in which the default certificate validation procedure is not adequate and some custom validation logic is required. This custom logic can be provided via the <System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> property. As an illustration, following sections provide some examples.

Reviewer: We may also mention [CertificateChainPolicy](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslserverauthenticationoptions.certificatechainpolicy?view=net-7.0#system-net-security-sslserverauthenticationoptions-certificatechainpolicy) that was added in 7.0 and perhaps show validation with custom trust.

```csharp
static bool CustomCertificateValidationCallback(
    object sender,
    X509Certificate? certificate,
    X509Chain? chain,
    SslPolicyErrors sslPolicyErrors)
{
    if (certificate == null)
    {
        return false;
    }

    const string ExpectedPublicKey =
        "3082010A0282010100C204ECF88CEE04C2B3D850D57058CC9318EB5C" +
        "A86849B022B5F9959EB12B2C763E6CC04B604C4CEAB2B4C00F80B6B0" +
        "F972C98602F95C415D132B7F71C44BBCE9942E5037A6671C618CF641" +
        "42C546D31687279F74EB0A9D11522621736C844C7955E4D16BE8063D" +
        "481552ADB328DBAAFF6EFF60954A776B39F124D131B6DD4DC0C4FC53" +
        "B96D42ADB57CFEAEF515D23348E72271C7C2147A6C28EA374ADFEA6C" +
        "B572B47E5AA216DC69B15744DB0A12ABDEC30F47745C4122E19AF91B" +
        "93E6AD2206292EB1BA491C0C279EA3FB8BF7407200AC9208D98C5784" +
        "538105CBE6FE6B5498402785C710BB7370EF6918410745557CF9643F" +
        "3D2CC3A97CEB931A4C86D1CA850203010001";

    return certificate.GetPublicKeyString().Equals(ExpectedPublicKey);
}
```

## Considerations for client certificate validation

> Server applications need to be careful when requiring and validating client certificates. Instead of sending the entire certificate chain, clients can configure the [AIA (Authority Information Access)](http://www.pkiglobe.org/auth_info_access.html) extension which specifies where the issuer certificate can be downloaded. The server will then download the issuer certificate when building the <xref:System.Security.Cryptography.X509Certificates.X509Chain> for the client certificate. Similarly, servers may need to contact external servers to ensure that the client certificate has not been revoked.

Reviewer: The AIA is not specific to client certs IMHO. But I understand the example bellow is mainly for servers.

## Custom X509Certificate validation

There are certain scenarios in which the default certificate validation procedure is not adequate and some custom validation logic is required. Parts of the validation logic can be customized by specifying <xref:System.Net.Security.SslClientAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType> or <xref:System.Net.Security.SslServerAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType> (available since .NET 7). Alternatively, completely custom logic can be provided via the <System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> property. As an illustration, following sections provide some examples.

### Custom Certificate Trust

When encountering a certificate which was not issued by any of the certificate authorities trusted by the machine (including self-signed certificates), the default certificate validation procedure will fail. One possible way to resolve this is add the necessary issuer certificates to the machine trusted store. That, however, might affect other applications on the system and is not always possible.

> The alternative solution is to specify custom trusted root certificates via an <xref:System.Security.Cryptography.X509Certificates.X509ChainPolicy>.

Reviewer 2: The policy specified below will replace the default/system trust list with an explicit trust list.  So while that might make a request to `somecomputer.localdomain` work, it will make a request to `api.github.com` fail (since it won't chain up to anything in the CustomTrustStore collection).  The sample as-is is probably fine, but the wording should be cleaned up to indicate that the first paragraph (add the custom root to system trust) and the second (use a custom trust list) are not equivalent options.

```csharp
SslClientAuthenticationOptions clientOptions = new();

clientOptions.CertificateChainPolicy = new X509ChainPolicy()
{
    TrustMode = X509ChainTrustMode.CustomRootTrust,
    CustomTrustStore = {
        customIssuerCert
    }
};
```

### Ignoring specific validation errors

Consider an IoT device without a persistent clock. After powering on, the clock of the device would start many years in the past and, therefore, all certificates would be considered "not yet valid". The following code shows a validation callback implementation which ignores validity period violations.

```csharp
static bool CustomCertificateValidationCallback(
    object sender,
    X509Certificate? certificate,
    X509Chain? chain,
    SslPolicyErrors sslPolicyErrors)
{

>     if (chain is null)
>     {
>         return false;
>     }
> 
>     foreach (X509ChainStatus status in chain.ChainStatus)
>     {
>         if ((status.Status & ~X509ChainStatusFlags.NotTimeValid) != X509ChainStatusFlags.NoError)
>         {
>             return false;
>         }
>     }
> 
>     return true;

Reviewer 2: This is a bad example, since it will allow host mismatch errors through.  A more robust version would be something like

```suggestion
    // Anything that would have been accepted by default is OK
    if (sslPolicyErrors == SslPolicyErrors.None)
    {
        return true;
    }
    
    // If there is something wrong other than a chain processing error, don't trust it.
    if (sslPolicyErrors != SslPolicyErrors.RemoteCertificateChainErrors)
    {
        return false;
    }
    
    Debug.Assert(chain is not null);
    
    foreach (X509ChainStatus status in chain.ChainStatus)
    {
        // If an error other than `NotTimeValid` (or `NoError`) is present, don't trust it.
        if ((status.Status & ~X509ChainStatusFlags.NotTimeValid) != X509ChainStatusFlags.NoError)
        {
            return false;
        }
    }

    return true;
```

}
```

### Certificate pinning

Another situation where custom certificate validation is necessary is when clients expect servers to use a specific certificate, or a certificate from a small set of known certificates. This practice is known as [certificate pinning](https://owasp.org/www-community/controls/Certificate_and_Public_Key_Pinning). The following code snippet shows a validation callback which checks that the server presents a certificate with a specific known public key.

```csharp
static bool CustomCertificateValidationCallback(
    object sender,
    X509Certificate? certificate,
    X509Chain? chain,
    SslPolicyErrors sslPolicyErrors)
{

>     if (certificate == null)

Reviewer 2: This sample implementation should probably still reject RemoteCertificateNameMismatch

---
title: TLS/SSL best practices

> description: Learn best practices to using SslStream in .NET

Reviewer 3: ```suggestion
description: Learn the best practices when using SslStream in .NET.
```

author: rzikm
ms.author: radekzikmund
ms.date: 10/25/2022
---

# TLS/SSL best practices

TLS (Transport Layer Security) is a cryptographic protocol designed to secure communication between two computers over the internet. The TLS protocol is exposed in .NET via the <xref:System.Net.Security.SslStream> class.

This article presents best practices for setting up secure communication between client and server and assumes use of .NET Core. For best practices for .NET Framework, see [Transport Layer Security (TLS) best practices with the .NET Framework](/dotnet/framework/network-programming/tls).

## Selecting TLS version

> While it is possible to specify the version of the TLS protocol to be used via the SslProtocols property, it is recommended to defer to the operating system settings by using <xref:System.Security.Authentication.SslProtocols.None> value (this is the default).

Author: It is actually `EnabledSslProtocols`, and there is one on `SslClientAuthenticationOptions` and `SslServerAuthenticationOptions`. should I xref it to one of these?

> While it is possible to specify the version of the TLS protocol to be used via the SslProtocols property, it is recommended to defer to the operating system settings by using <xref:System.Security.Authentication.SslProtocols.None> value (this is the default).

Reviewer 3: Yes, please.

- By passing a <xref:System.Net.Security.SslStreamCertificateContext> in the <xref:System.Net.Security.SslServerAuthenticationOptions.ServerCertificateContext?displayProperty=nameWithType> property

The recommended approach is to use the <xref:System.Net.Security.SslServerAuthenticationOptions.ServerCertificateContext?displayProperty=nameWithType> property. When the certificate is obtained by one of the other two ways, a <xref:System.Net.Security.SslStreamCertificateContext> instance is created internally by the <xref:System.Net.Security.SslStream> implementation. Creating a <xref:System.Net.Security.SslStreamCertificateContext> involves building an <xref:System.Security.Cryptography.X509Certificates.X509Chain> which is a CPU intensive operation. It is more efficient to create a <xref:System.Net.Security.SslStreamCertificateContext> once and reuse it for multiple <xref:System.Net.Security.SslStream> instances.

Reusing <xref:System.Net.Security.SslStreamCertificateContext> instances also enables additional features such us [TLS session resumption](https://datatracker.ietf.org/doc/html/rfc5077) on Linux servers.

> ## Custom X509Certificate validation

Reviewer 3: ```suggestion
## Custom `X509Certificate` validation
```

> There are certain scenarios in which the default certificate validation procedure is not adequate and some custom validation logic is required. Parts of the validation logic can be customized by specifying <xref:System.Net.Security.SslClientAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType> or <xref:System.Net.Security.SslServerAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType> (available since .NET 7). Alternatively, completely custom logic can be provided via the <System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> property. As an illustration, following sections provide some examples.

Reviewer 3: ```suggestion
There are certain scenarios in which the default certificate validation procedure isn't adequate and some custom validation logic is required. Parts of the validation logic can be customized by specifying <xref:System.Net.Security.SslClientAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType> or <xref:System.Net.Security.SslServerAuthenticationOptions.CertificateChainPolicy?displayProperty=nameWithType>. Alternatively, completely custom logic can be provided via the <System.Net.Security.SslClientAuthenticationOptions.RemoteCertificateValidationCallback> property. For more information, see [Custom certificate trust](#custom-certificate-trust).
```

> ### Custom Certificate Trust

Reviewer 3: ```suggestion
### Custom certificate trust
```

> When encountering a certificate which was not issued by any of the certificate authorities trusted by the machine (including self-signed certificates), the default certificate validation procedure will fail. One possible way to resolve this is add the necessary issuer certificates to the machine trusted store. That, however, might affect other applications on the system and is not always possible.

Reviewer 3: ```suggestion
When encountering a certificate that wasn't issued by any of the certificate authorities trusted by the machine (including self-signed certificates), the default certificate validation procedure will fail. One possible way to resolve this is to add the necessary issuer certificates to the machine's trusted store. That, however, might affect other applications on the system and is not always possible.
```

> The alternative solution is to specify custom trusted root certificates via an <xref:System.Security.Cryptography.X509Certificates.X509ChainPolicy>. The example below specifies custom trust list which will be used instead of the system trust list during validation.

Reviewer 3: ```suggestion
The alternative solution is to specify custom trusted root certificates via an <xref:System.Security.Cryptography.X509Certificates.X509ChainPolicy>. To specify a custom trust list that will be used instead of the system trust list during validation, consider the following example:
```

```csharp
SslClientAuthenticationOptions clientOptions = new();

clientOptions.CertificateChainPolicy = new X509ChainPolicy()
{
    TrustMode = X509ChainTrustMode.CustomRootTrust,

>     CustomTrustStore = {
>         customIssuerCert
>     }

Reviewer 3: ```suggestion
    CustomTrustStore =
    {
        customIssuerCert
    }
```

};
```

> Note that clients configured with the policy above would accept *only* certificates trusted by `customIssuerCert`.

Reviewer 3: ```suggestion
Clients configured with the preceding policy would only accept certificates trusted by `customIssuerCert`.
```

> ### Ignoring specific validation errors

Reviewer 3: ```suggestion
### Ignore specific validation errors
```

> Consider an IoT device without a persistent clock. After powering on, the clock of the device would start many years in the past and, therefore, all certificates would be considered "not yet valid". The following code shows a validation callback implementation which ignores validity period violations.

Reviewer 3: ```suggestion
Consider an IoT device without a persistent clock. After powering on, the clock of the device would start many years in the past and, therefore, all certificates would be considered "not yet valid". Consider the following code that shows a validation callback implementation ignoring validity period violations.
```

    
    foreach (X509ChainStatus status in chain.ChainStatus)
    {
        // If an error other than `NotTimeValid` (or `NoError`) is present, don't trust it.
        if ((status.Status & ~X509ChainStatusFlags.NotTimeValid) != X509ChainStatusFlags.NoError)
        {
            return false;
        }
    }

    return true;
}
```

### Certificate pinning

Another situation where custom certificate validation is necessary is when clients expect servers to use a specific certificate, or a certificate from a small set of known certificates. This practice is known as [certificate pinning](https://owasp.org/www-community/controls/Certificate_and_Public_Key_Pinning). The following code snippet shows a validation callback which checks that the server presents a certificate with a specific known public key.

```csharp
static bool CustomCertificateValidationCallback(
    object sender,
    X509Certificate? certificate,
    X509Chain? chain,
    SslPolicyErrors sslPolicyErrors)
{
    // If there is something wrong other than a chain processing error, don't trust it.
    if (sslPolicyErrors != SslPolicyErrors.RemoteCertificateChainErrors)
    {
        return false;

>     }

Reviewer 2: This block would make sense if the caller expects the certificate to not be trusted.  But if it was trusted, why would that make the pinning fail?

In this particular scenario, it'd be something like `if ((sslPolicyErrors & ~SslPolicyErrors.RemoteCertificateChainErrors) != 0)` (ignoring the chain errors bit, fail if any (other) bit is set).... which only allows None or RemoteCertificateChainErrors through.

        "42C546D31687279F74EB0A9D11522621736C844C7955E4D16BE8063D" +
        "481552ADB328DBAAFF6EFF60954A776B39F124D131B6DD4DC0C4FC53" +
        "B96D42ADB57CFEAEF515D23348E72271C7C2147A6C28EA374ADFEA6C" +
        "B572B47E5AA216DC69B15744DB0A12ABDEC30F47745C4122E19AF91B" +
        "93E6AD2206292EB1BA491C0C279EA3FB8BF7407200AC9208D98C5784" +
        "538105CBE6FE6B5498402785C710BB7370EF6918410745557CF9643F" +
        "3D2CC3A97CEB931A4C86D1CA850203010001";

    return certificate.GetPublicKeyString().Equals(ExpectedPublicKey);
}
```

## Considerations for client certificate validation

Server applications need to be careful when requiring and validating client certificates. Certificates may contain the [AIA (Authority Information Access)](http://www.pkiglobe.org/auth_info_access.html) extension which specifies where the issuer certificate can be downloaded. The server may therefore attempt to download the issuer certificate from external server when building the <xref:System.Security.Cryptography.X509Certificates.X509Chain> for the client certificate. Similarly, servers may need to contact external servers to ensure that the client certificate has not been revoked.

> The need to contact external servers when building and validating the <xref:System.Security.Cryptography.X509Certificates.X509Chain> may expose the application to denial of service attacks if the external servers are slow to respond. Therefore, server applications should configure the <xref:System.Security.Cryptography.X509Certificates.X509Chain> building behavior using the <xref:System.Net.Security.SslServerAuthenticationOptions.CertificateChainPolicy> (available since .NET 7).

Reviewer 3: ```suggestion
The need to contact external servers when building and validating the <xref:System.Security.Cryptography.X509Certificates.X509Chain> may expose the application to denial of service attacks if the external servers are slow to respond. Therefore, server applications should configure the <xref:System.Security.Cryptography.X509Certificates.X509Chain> building behavior using the <xref:System.Net.Security.SslServerAuthenticationOptions.CertificateChainPolicy>.
```

---
title: TLS/SSL Best Practices
description: Learn best practices to using SslStream in .NET
author: rzikm
ms.author: rzikm
ms.date: 10/25/2022
---

# TLS/SSL Best Practices

[TLS (Transport Layer Security)](https://en.wikipedia.org/wiki/Transport_Layer_Security) is a cryptographic protocol designed to secure communication between two computers over the internet. The TLS protocol is exposed in .NET via the [`SslStream`](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslstream) class.

> This article presents best practices for setting up secure communication between client and server and assumes .NET Core version 3.1 or later. For best practices for .NET Framework, see [Transport Layer Security (TLS) best practices with the .NET Framework](https://learn.microsoft.com/en-us/dotnet/framework/network-programming/tls).

Reviewer 3: We cannot have links in them that contain the locale (i.e.; `/en-us/`), nor can we use fully qualified URLs. Please install and use the Learn Authoring Pack tool for Visual Studio Code to leverage the tooling support for issues like this. It can create links that are site relative with ease, and can be used to insert `xref`. That's the other concern with this PR, code references should use the `xref` syntax rather than the MD link syntax.

https://review.learn.microsoft.com/en-us/help/platform/docs-authoring-pack-overview?branch=main
